pkg/report/report.go

```diff
@@ -92,168 +89,93 @@ func runQuery(ctx context.Context, tx *sqlx.Tx, prom PromQuerier, query db.Query
+		if err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("failed to process sample: %w", err))
+		} else {
+			records = append(records, *record)
 		}
 	}
 
-	return nil
+	return multierr.Append(errs, odooClient.SendData(ctx, records))
 }
 
-func processSample(ctx context.Context, tx *sqlx.Tx, ts time.Time, query db.Query, s *model.Sample) error {
-	category, err := getMetricLabel(s.Metric, "category")
-	if err != nil {
-		return err
-	}
-	productLabel, err := getMetricLabel(s.Metric, "product")
-	if err != nil {
-		return err
-	}
+func processSample(ctx context.Context, odooClient OdooClient, args ReportArgs, from time.Time, s *model.Sample) (*odoo.OdooMeteredBillingRecord, error) {
+	metricLabels := s.Metric
 
-	skey, err := sourcekey.Parse(string(productLabel))
-	if err != nil {
-		return fmt.Errorf("failed to parse source key from product label: %w", err)
+	salesOrderID := ""
+	if args.OverrideSalesOrderID != "" {
+		salesOrderID = args.OverrideSalesOrderID
+	} else {
+		sid, err := getMetricLabel(s.Metric, SalesOrderIDLabel)
+		if err != nil {
+			return nil, err
+		}
+		salesOrderID = string(sid)
 	}
 
-	var upsertedTenant db.Tenant
-	err = upsertTenant(ctx, tx, &upsertedTenant, db.Tenant{
-		Source: skey.Tenant(),
-	}, ts)
+	labelList, err := json.Marshal(metricLabels)
 	if err != nil {
-		return fmt.Errorf("failed to upsert tenant '%s': %w", skey.Tenant(), err)
-	}
-
-	var upsertedCategory db.Category
-	if err := upsertCategory(ctx, tx, &upsertedCategory, db.Category{Source: string(category)}); err != nil {
-		return err
-	}
-
-	sourceLookup := skey.LookupKeys()
-
-	var product db.Product
-	if err := getBySourceKeyAndTime(ctx, tx, &product, pgx.Identifier{"products"}, sourceLookup, ts); err != nil {
-		return fmt.Errorf("failed to load product for '%s': %w", productLabel, err)
+		return nil, err
 	}
 
-	var discount db.Discount
-	if err := getBySourceKeyAndTime(ctx, tx, &discount, pgx.Identifier{"discounts"}, sourceLookup, ts); err != nil {
-		return fmt.Errorf("failed to load discount for '%s': %w", productLabel, err)
-	}
+	vm := jsonnet.MakeVM()
+	vm.ExtCode("labels", string(labelList))
 
-	var upsertedDateTime db.DateTime
-	err = upsertDateTime(ctx, tx, &upsertedDateTime, db.BuildDateTime(ts))
+	instance, err := vm.EvaluateAnonymousSnippet("instance.json", args.InstanceJsonnet)
 	if err != nil {
-		return fmt.Errorf("failed to upsert date_time '%s': %w", ts.Format(time.RFC3339), err)
+		return nil, err
 	}
-
-	var upsertedFact db.Fact
-	err = upsertFact(ctx, tx, &upsertedFact, db.Fact{
-		DateTimeId: upsertedDateTime.Id,
-		TenantId:   upsertedTenant.Id,
-		CategoryId: upsertedCategory.Id,
-		QueryId:    query.Id,
-		ProductId:  product.Id,
-		DiscountId: discount.Id,
-		Quantity:   float64(s.Value),
-	})
+	instanceStr := ""
+	err = json.Unmarshal([]byte(instance), &instanceStr)
 	if err != nil {
-		return fmt.Errorf("failed to upsert fact '%s': %w", ts.Format(time.RFC3339), err)
+		return nil, err
```

Comment:
Here some context to the error could help the end user of the tool

```suggestion
		return nil, fmt.Errorf("failed to interpolate instance template: %w", err)
```

Reply:
Good call! I've already started forking the repo so I'll implement that in the new repo.